In [13]:
import pandas as pd
import numpy as np

In [20]:
airlines = pd.read_csv('https://raw.githubusercontent.com/tidyverse/nycflights13/master/data-raw/airlines.csv')
airports = pd.read_csv('https://raw.githubusercontent.com/tidyverse/nycflights13/master/data-raw/airports.csv')
planes = pd.read_csv('https://raw.githubusercontent.com/tidyverse/nycflights13/master/data-raw/planes.csv')
weather = pd.read_csv('https://raw.githubusercontent.com/tidyverse/nycflights13/master/data-raw/weather.csv')


Your task in this week’s assignment is to answer three questions: 
1. What is the northernmost airport in the United States? 
2. What is the easternmost airport in the United States? 
3. On February 12th, 2013, which New York area airport had the windiest weather? 

#### What is the northernmost airport in the United States?

In [128]:
airports.head()

faa                           name        lat        lon   alt  tz dst  \
0  04G              Lansdowne Airport  41.130472 -80.619583  1044  -5   A   
1  06A  Moton Field Municipal Airport  32.460572 -85.680028   264  -6   A   
2  06C            Schaumburg Regional  41.989341 -88.101243   801  -6   A   
3  06N                Randall Airport  41.431912 -74.391561   523  -5   A   
4  09J          Jekyll Island Airport  31.074472 -81.427778    11  -5   A   

              tzone  
0  America/New_York  
1   America/Chicago  
2   America/Chicago  
3  America/New_York  
4  America/New_York

Latitude is expressed in degrees above or below the equator, positive values being north and negative values being south.

In [140]:
airports.sort_values(by='lat', ascending=False).head()

faa                                        name        lat         lon  \
417  EEN                     Dillant Hopkins Airport  72.270833   42.898333   
230  BRW                  Wiley Post Will Rogers Mem  71.285446 -156.766003   
110  AIN                          Wainwright Airport  70.638056 -159.994722   
708  K03                               Wainwright As  70.613378 -159.860350   
152  ATK  Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000   

     alt  tz dst              tzone  
417  149  -5   A                NaN  
230   44  -9   A  America/Anchorage  
110   41  -9   A  America/Anchorage  
708   35  -9   A  America/Anchorage  
152   96  -9   A  America/Anchorage

We see that sorting shows up EEN or Dillant Hopkins Airport being the most northern. The other way we could look at this is by using the .max() method. While doing this we will need to account of NaN values. We can do this with a dropna method.

In [141]:
airports.dropna(how='all').lat.max()

72.270833

we see that the max latitude is ~72.27. We can use a simple comparision to ensure this is our candidate.

In [143]:
airports[(airports.lat) > 72]

faa                     name        lat        lon  alt  tz dst tzone
417  EEN  Dillant Hopkins Airport  72.270833  42.898333  149  -5   A   NaN

Our answer is EEN or Dilliant Hopkins Airport.

___

#### What is the easternmost airport in the United States? 

Longitude is expressed in degrees relative to the Prime Meridian, with positive degrees being eastern and negative degrees being western. In the case of our answer, it lies so close to the Prime Meridian that it ends up having a very high positive longitude and therefore takes 1st place.

In [157]:
airports.sort_values(by='lon',ascending=False).head()

faa                           name        lat         lon   alt  tz dst  \
1290  SYA                   Eareckson As  52.712275  174.113620    98  -9   A   
942   MYF               Montgomery Field  32.475900  117.759000    17   8   A   
396   DVT  Deer Valley Municipal Airport  33.411700  112.457000  1478   8   A   
417   EEN        Dillant Hopkins Airport  72.270833   42.898333   149  -5   A   
444   EPM     Eastport Municipal Airport  44.910111  -67.012694    45  -5   A   

                  tzone  
1290  America/Anchorage  
942      Asia/Chongqing  
396      Asia/Chongqing  
417                 NaN  
444    America/New_York

Similar to what we did above, we first need to remove any NaN data that is relevant to our search.

In [158]:
airports.dropna(how='all').lon.max()

174.11362

In [159]:
airports[(airports.lon) > 174]

faa          name        lat        lon  alt  tz dst              tzone
1290  SYA  Eareckson As  52.712275  174.11362   98  -9   A  America/Anchorage

We see that searching for the max value for the desired category and comparing it to our result find a correct answer.

___

#### On February 12th, 2013, which New York area airport had the windiest weather? 

In [112]:
import datetime

In [113]:
windgust = weather[(weather.time_hour>='2013-02-12') & (weather.time_hour<'2013-02-13')]

In [114]:
windgust.dropna(subset=['wind_speed', 'wind_gust'],how='any', inplace=True)

/tmp/ipykernel_7666/3362276890.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  windgust.dropna(subset=['wind_speed', 'wind_gust'],how='any', inplace=True)


In [115]:
windgust[windgust.wind_speed == windgust['wind_speed'].max()]

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
18417    LGA  2013      2   12     2  42.98  26.06  50.94     290.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
18417     23.0156   31.07106     0.0    1007.1   10.0  2013-02-12T07:00:00Z

The answer is LGA or Laguardia Airport